### PR TITLE
baloo: patch to support btrfs better

### PR DESCRIPTION
**Pull request by peterhoeg**
###### Description of changes

baloo suffers from issues when running on btrfs as well as with certain LVM/dm-crypt setups
where the device id will change on reboot causing baloo to reindex all the files and then having
duplicate files. A patch has been proposed that addresses this, which has not been accepted yet.
However, without this patch, people tend to disable baloo rather than dealing with the constant
reindexing.

https://bugs.kde.org/show_bug.cgi?id=402154

I have been running with the patch applied for a quite a while and it makes a *huge* difference.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by peterhoeg** [comment]
Cc: @NixOS/qt-kde 

Just to add a little context (and because I'm tired of waiting for a number of KDE bits to recompile on updates), I think we should move ahead with merging this despite it not being upstream yet.

Pros - the experience out of the box for users on btrfs is *much* better:
 - no constant reindexing 
 - no duplicate results
 - no phantom results left behind after the file is deleted

Cons:
 - there might be some corner cases where baloo isn't indexing. One example given is mounting 2 file system images with similar FS id. I would argue that that is so niche so as to not matter

In conclusion, this is a definite quality of life improvement.
   -

**Comment by peterhoeg** [comment]
Unless I hear any strong objections, I'll be merging this at the end of the week.

**Comment by peterhoeg** [comment]
Cc: @K900 @NickCao

**Review by NickCao — APPROVED** [review]
Not a KDE (baloo) user myself, but the patch looks reasonable.

**Comment by K900** [comment]
I don't mind the patch, but I also don't really use Baloo that much so it's hard for me to estimate the impact.

**Comment by peterhoeg** [comment]
The impact is really “make baloo usable again on btrfs”.

**Comment by K900** [comment]
Yeah, I see the btrfs improvements, I'm just wondering if this has any unexpected side effects.

**Comment by alois31** [comment]
My machine has performed two reindexes since this patch was included in my system configuration. Is that intended behavior (i.e. will the reindexes eventually stop again)?

**Comment by peterhoeg** [comment]
Once is expected, not two. Do you mind trying to do a manual disable,
purge, index cycle?

What FS are you offering?

**Comment by alois31** [comment]
All indexed directories are on btrfs. I have tried disabling, purging and re-enabling, and after waiting for the indexing to happen and two reboots I am happy to report that no pointless reindexing seems to be happening any more, and on top the index is much smaller now because it does not contain the several copies of the old entries any more. Maybe recommending this in the release notes would be a good idea?

**Comment by peterhoeg** [comment]
Good point. Will do.
